### ISSUE TITLE
Should we store the history stack in project files?

### ISSUE THREAD

**Issue by ferdnyc** [OpenShot/openshot-qt#1923]
Bug #1768 deals with an issue loading project data into OpenShot, and while it's not yet _certain_ exactly what caused the issue, one of the things I noticed in examining it was the `"history": { ... }` data included in the project data.

In that particular case, the saved history data alone accounted for more than _half_ of the entire (1.4 MB) project data.

Even examining my own, much simpler test project files, the history sections occupy...
* 285/464 formatted lines (61.5% by line count)
* 5929/9197 formatted lines (64.5% by line count)
* 2020/2586 formatted lines (**78% (!!)** by line count)

Which leads to the obvious question: Should we be flushing the undo/redo history on save? (Or, if not flushing it, then at least omitting it from the save file. It can stay in memory until the end of the session.)

It's worth noting that, currently, the undo/redo history **IS** restored on project load, and can be accessed just as if the new session was a continuation of the old one. So, arguably there's some value to having the history stored in the save file. 

However, it's worth at least asking the question of whether that's really _useful_ in practice, and whether any users really do ever load a project and then immediately Undo or Redo any actions from before saving. I only have my own anecdotal opinion to go on, but I don't think that would ever even have _occurred_ to me. I've certainly never actually done it.

**Comment by ferdnyc** [comment]
An additional dimension on which to consider this question: 

In the discussion at PR #2275, I mentioned that the absolute paths stored in the project history for _internal_ assets causes problems when running OpenShot from the AppImage builds, because the path to OpenShot's own files isn't static. (The AppImage is mounted to a directory `/tmp/.mount_XXXXXX/` at launch, with the characters after `mount_` changing each run.)

I just managed to **crash** OpenShot 2.4.3 running from my installed `/usr/bin/openshot`, due to this issue.

I loaded up a saved project file, hit Undo, hit Redo, and \*bang\*:
```
 main_window:INFO actionRedo_trigger
terminate called after throwing an instance of 'openshot::InvalidFile'
  what():  File could not be opened.
Caught signal 6 (SIGABRT)
---- Unhandled Exception: Stack Trace ----
  /lib64/libc.so.6               ( abort                                     + 0x123 )  [0x7f83f4a2e5b9]
  /lib64/libstdc++.so.6          (                                           + 0x8ca9b)  [0x7f83f3491a9b]
  /lib64/libstdc++.so.6          (                                           + 0x92efc)  [0x7f83f3497efc]
  /lib64/libstdc++.so.6          (                                           + 0x91f19)  [0x7f83f3496f19]
  /lib64/libstdc++.so.6          ( __gxx_personality_v0                      + 0x2e8 )  [0x7f83f34978c8]
  /lib64/libgcc_s.so.1           (                                           + 0x107b3)  [0x7f83f31fd7b3]
  /lib64/libgcc_s.so.1           ( _Unwind_Resume                            + 0x12d )  [0x7f83f31fe0cd]
  /lib64/libopenshot.so.16       ( openshot::QtImageReader::QtImageReader(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)  + 0xe7  )  [0x7f83cd53b5f7]
  /lib64/libopenshot.so.16       ( openshot::Mask::SetJsonValue(Json::Value)  + 0x55d )  [0x7f83cd4eaeb5]
  /lib64/libopenshot.so.16       ( openshot::Timeline::apply_json_to_effects(Json::Value, openshot::EffectBase*)  + 0x345 )  [0x7f83cd5497c3]
  /lib64/libopenshot.so.16       ( openshot::Timeline::apply_json_to_effects(Json::Value)  + 0x292 )  [0x7f83cd549400]
  /lib64/libopenshot.so.16       ( openshot::Timeline::ApplyJsonDiff(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)  + 0x2c3 )  [0x7f83cd547faf]
  /usr/lib64/python3.6/site-packages/_openshot.so (                                           + 0x12711e)  [0x7f83cde9811e]
  /lib64/libpython3.6m.so.1.0    ( _PyCFunction_FastCallDict                 + 0x92  )  [0x7f83f58f9942]
  /lib64/libpython3.6m.so.1.0    (                                           + 0x182bdd)  [0x7f83f5908bdd]
  /lib64/libpython3.6m.so.1.0    ( _PyEval_EvalFrameDefault                  + 0x34a )  [0x7f83f593912a]
  /lib64/libpython3.6m.so.1.0    (                                           + 0x13ad5a)  [0x7f83f58c0d5a]
  /lib64/libpython3.6m.so.1.0    (                                           + 0x182cc2)  [0x7f83f5908cc2]
  /lib64/libpython3.6m.so.1.0    ( _PyEval_EvalFrameDefault                  + 0x34a )  [0x7f83f593912a]
  /lib64/libpython3.6m.so.1.0    (                                           + 0x13ad5a)  [0x7f83f58c0d5a]
  /lib64/libpython3.6m.so.1.0    (                                           + 0x182cc2)  [0x7f83f5908cc2]
  /lib64/libpython3.6m.so.1.0    ( _PyEval_EvalFrameDefault                  + 0x34a )  [0x7f83f593912a]
  /lib64/libpython3.6m.so.1.0    (                                           + 0x13ad5a)  [0x7f83f58c0d5a]
  /lib64/libpython3.6m.so.1.0    (                                           + 0x182cc2)  [0x7f83f5908cc2]
  /lib64/libpython3.6m.so.1.0    ( _PyEval_EvalFrameDefault                  + 0x34a )  [0x7f83f593912a]
  /lib64/libpython3.6m.so.1.0    (                                           + 0x13ad5a)  [0x7f83f58c0d5a]
  /lib64/libpython3.6m.so.1.0    (                                           + 0x182cc2)  [0x7f83f5908cc2]
  /lib64/libpython3.6m.so.1.0    ( _PyEval_EvalFrameDefault                  + 0x34a )  [0x7f83f593912a]
  /lib64/libpython3.6m.so.1.0    ( _PyFunction_FastCallDict                  + 0x12a )  [0x7f83f588b0fa]
  /lib64/libpython3.6m.so.1.0    ( _PyObject_FastCallDict                    + 0x31e )  [0x7f83f588baae]
  /lib64/libpython3.6m.so.1.0    ( _PyObject_Call_Prepend                    + 0x60  )  [0x7f83f58973f0]
  /lib64/libpython3.6m.so.1.0    ( PyObject_Call                             + 0x43  )  [0x7f83f588be83]
  /usr/lib64/python3.6/site-packages/PyQt5/QtCore.so (                                           + 0x1ffdc0)  [0x7f83f3ed3dc0]
  /usr/lib64/python3.6/site-packages/PyQt5/QtCore.so (                                           + 0x200298)  [0x7f83f3ed4298]
  /usr/lib64/python3.6/site-packages/PyQt5/QtCore.so (                                           + 0x200590)  [0x7f83f3ed4590]
  /usr/lib64/python3.6/site-packages/PyQt5/QtCore.so (                                           + 0x200f47)  [0x7f83f3ed4f47]
  /lib64/libQt5Core.so.5         ( QMetaObject::activate(QObject*, int, int, void**)  + 0x6f7 )  [0x7f83f3a15ec7]
  /lib64/libQt5Widgets.so.5      ( QAction::triggered(bool)                  + 0x46  )  [0x7f83ead2fc26]
  /lib64/libQt5Widgets.so.5      ( QAction::activate(QAction::ActionEvent)   + 0x100 )  [0x7f83ead322f0]
  /lib64/libQt5Widgets.so.5      (                                           + 0x252f01)  [0x7f83eae1ff01]
  /lib64/libQt5Widgets.so.5      ( QAbstractButton::mouseReleaseEvent(QMouseEvent*)  + 0xe5  )  [0x7f83eae20135]
  /lib64/libQt5Widgets.so.5      ( QToolButton::mouseReleaseEvent(QMouseEvent*)  + 0xe   )  [0x7f83eaf0dffe]
  /lib64/libQt5Widgets.so.5      ( QWidget::event(QEvent*)                   + 0x1d8 )  [0x7f83ead755c8]
  /lib64/libQt5Widgets.so.5      ( QToolButton::event(QEvent*)               + 0x37  )  [0x7f83eaf0e0a7]
  /lib64/libQt5Widgets.so.5      ( QApplicationPrivate::notify_helper(QObject*, QEvent*)  + 0x85  )  [0x7f83ead36475]
  /lib64/libQt5Widgets.so.5      ( QApplication::notify(QObject*, QEvent*)   + 0x458 )  [0x7f83ead3ddd8]
  /usr/lib64/python3.6/site-packages/PyQt5/QtWidgets.so (                                           + 0x16723e)  [0x7f83eb3af23e]
  /lib64/libQt5Core.so.5         ( QCoreApplication::notifyInternal2(QObject*, QEvent*)  + 0x86  )  [0x7f83f39edc36]
  /lib64/libQt5Widgets.so.5      ( QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool)  + 0x1ad )  [0x7f83ead3d0cd]
  /lib64/libQt5Widgets.so.5      (                                           + 0x1c3368)  [0x7f83ead90368]
  /lib64/libQt5Widgets.so.5      (                                           + 0x1c5f0e)  [0x7f83ead92f0e]
  /lib64/libQt5Widgets.so.5      ( QApplicationPrivate::notify_helper(QObject*, QEvent*)  + 0x85  )  [0x7f83ead36475]
  /lib64/libQt5Widgets.so.5      ( QApplication::notify(QObject*, QEvent*)   + 0x210 )  [0x7f83ead3db90]
  /usr/lib64/python3.6/site-packages/PyQt5/QtWidgets.so (                                           + 0x16723e)  [0x7f83eb3af23e]
  /lib64/libQt5Core.so.5         ( QCoreApplication::notifyInternal2(QObject*, QEvent*)  + 0x86  )  [0x7f83f39edc36]
  /lib64/libQt5Gui.so.5          ( QGuiApplicationPrivate::processMouseEvent(QWindowSystemInterfacePrivate::MouseEvent*)  + 0x6c3 )  [0x7f83ea7e3f43]
  /lib64/libQt5Gui.so.5          ( QGuiApplicationPrivate::processWindowSystemEvent(QWindowSystemInterfacePrivate::WindowSystemEvent*)  + 0x135 )  [0x7f83ea7e6085]
  /lib64/libQt5Gui.so.5          ( QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>)  + 0xab  )  [0x7f83ea7c146b]
  /lib64/libQt5XcbQpa.so.5       (                                           + 0xdb7cf)  [0x7f83bda237cf]
  /lib64/libQt5Core.so.5         ( QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>)  + 0x13b )  [0x7f83f39ecb7b]
---- End of Stack Trace ----
```

Sure enough, the last item in the history for that project was the addition of a transition from an AppImage session:
```json
         {
            "type" : "insert",
            "old_values" : null,
            "value" : {
               "start" : 0,
               "id" : "KIA1GT0JVJ",
               "title" : "Transition",
               "replace_image" : false,
               "reader" : {
                  "has_audio" : false,
                  "acodec" : "",
                  "interlaced_frame" : false,
                  "height" : 576,
                  "sample_rate" : 0,
                  "pixel_ratio" : {
                     "den" : 1,
                     "num" : 1
                  },
                  "channel_layout" : 4,
                  "video_timebase" : {
                     "num" : 1,
                     "den" : 30
                  },
                  "duration" : 86400,
                  "path" : "/tmp/.mount_Q6SNRa/usr/bin/transitions/common/fade.svg",
                  "has_single_image" : true,
                  "top_field_first" : true,
                  "type" : "QtImageReader",
                  "vcodec" : "",
                  "channels" : 0,
                  "width" : 720,
                  "file_size" : "1658880",
                  "fps" : {
                     "num" : 30,
                     "den" : 1
                  },
    [...snip...]
            "key" : [
               "effects"
            ],
            "partial" : false
         }
      ],
```

Unlike the project file loader, OpenShot's history manager doesn't do any checks on the paths being applied, and has absolutely no way to handle path invalidation. That makes it _particularly_ dangerous to be storing history across sessions in the project file data.

**Comment by peanutbutterandcrackers** [comment]
Just something I experienced just now:

I had a project I'd been working on for a few days now. It was 12 MB. Loading it and doing further edits were taking a lot of time. It was lagging the computer. Since I remembered this, I went ahead and deleted the entire contents of the `undo` part under `history` in the `.osp` file and the `.osp` file is now down to 115K (without spaces) [130K after re-saving it from OpenShot, with the original spacing and all] and it seems that I will be able to make further changes to the project.

**Comment by ferdnyc** [comment]
Huh. Now, that's _extremely_ interesting.

Conventional Wisdom™ has been that since the history stack is limited in size (though I forget what the limit is), there should be a maximum to the impact it can have on performance. But if that limit is too high, or not working correctly when the history is loaded from the project file, it could be that the undo history is actually responsible for a lot of the unresponsiveness people have complained about with OpenShot.

Additionally, other Conventional Wisdom™ (just generally when it comes to computer software) has always been that if software is experiencing performance problems, you can always save off your work, quit the program, and then restart it to get things back to a more responsive state. But by storing history in the project file, OpenShot may be sabotaging that exact solution to any performance issues experienced.

@jonoomph ? @DylanC ?

**Comment by ferdnyc** [comment]
Also, 12MB → 130K is patently insane, and making me question whether there really _is_ a limit being imposed on the history stack.

**Comment by peanutbutterandcrackers** [comment]
I just went and double checked, to be sure. And history limit is set to 20 (the default) in the preferences.

**Comment by DylanC** [comment]
@ferdnyc @peanutbutterandcrackers  From my testing it _seems_ to limit the undo history to 20, however, each entry on the undo list looks massive.

![undo_hist](https://user-images.githubusercontent.com/4488048/47970717-130fb680-e081-11e8-801c-c8db4a571049.png)

> It's plausible that it was / there were only 20. One of the issues with the history is that it stores the entire before/after state of the project data for each change. So as the project gets more complex, *each* history item grows larger and larger.

I think this estimation is probably close to the answer or the answer to some of the performance issues we have.

**Comment by peanutbutterandcrackers** [comment]
@DylanC - Which text editor are you using, Cap'ain? I was unable to look at the history properly. I deleted them using `d%dd` on `vi` but can't seem figure out how to paste the deleted part into another file, yet. Any ideas how I could do it (preferably in vi itself)?

This project uses a lot of keyframes so I was thinking about making a github repo for this project sometime later (once it gets proper shape) so that others can test OpenShot and play around with it using just it.

**Comment by ferdnyc** [comment]
> @DylanC - Which text editor are you using, Cap'ain? i
Yeah, that folding's actually _way_ nicer than Atom or anything else I've seen.

> I was unable to look at the history properly. I deleted them using `d%dd` on `vi` but can't seem figure out how to paste the deleted part into another file, yet. Any ideas how I could do it (preferably in vi itself)?

In vim (probably straight vi, too), if you've deleted a bunch of lines all in one go (they're in the cut buffer), you can use `:new` to open a window to an empty file (it'll split the screen, the new window above your old), then just hit `p` to put the deleted text into the new buffer. After that, just `:w filename` to save the new buffer to a file, as normal.

(Actually, even if it wasn't all in one go, `"2p` through `"9p` (yes, including the double quote) will put the contents of your 8 _prior_ deletions. `"1p` is technically a synonym for just `p`.)

> This project uses a lot of keyframes so I was thinking about making a github repo for this project sometime later (once it gets proper shape) so that others can test OpenShot and play around with it using just it.

Good (especially, _complex_) test/demo files are **sorely** needed, for debugging and testing purposes even more than for demonstration/dog-and-pony type uses. I could see things like that being part of the _main_ repo at some point. (Though ideally they'd be left out of packaging and bundling, or packaged separately if at all.)

> > It's plausible that it was / there were only 20. One of the issues with the history is that it stores the entire before/after state of the project data for each change. So as the project gets more complex, _each_ history item grows larger and larger.
> 
> I think this estimation is probably close to the answer or the answer to some of the performance issues we have.

If nothing else, it's becoming **painfully** clear that we at least need to offer the _option_ to exclude history from project files. (Either as a preference, or ideally, a per-project `Save As...` type option. Though the UI to implement that feels like it couldn't possibly escape being clunky, sigh.) Assuming we don't just make the decision — or, convince @jonoomph  to make the decision, really — to just do that unilaterally for all projects.

**Comment by ferdnyc** [comment]
(Oh, the thought that was meant to _precede_ my final paragraph about removing history from the project files was...)

> > It's plausible that it was / there were only 20. One of the issues with the history is that it stores the entire before/after state of the project data for each change. So as the project gets more complex, _each_ history item grows larger and larger.
> 
> I think this estimation is probably close to the answer or the answer to some of the performance issues we have.

Yeah, the efficiency of the history really breaks down as project complexity grows... in fact, the growth of the history's "weight" is probably something close to exponential.

Medium-term, it might be possible to find a good library/implementation for a JSON "diff" algorithm, and populate the history with only the _changes_ from one project state to another, instead of storing full snapshots. 

This also just reminded me of OpenShot/libopenshot#142 ­— if `json_cpp` isn't up to the task of more efficient history management, maybe `QJsonObject` is?

**Comment by DylanC** [comment]
> Which text editor are you using, Cap'ain?

> Yeah, that folding's actually way nicer than Atom or anything else I've seen.

My main editor is Sublime Text 3 but funnily enough the screenshot is just from an online json parser. 😆   http://json.parser.online.fr/

Btw, some inside knowledge from Peter on the Undo steps. If you do a split clip and "keep both sides" it will have 3 undo steps. This should be investigated at some point to see could we have less undo steps for this task.

**Comment by ferdnyc** [comment]
> Once your backwards-compatible thumbnail PR is done and merged,

Yeah, I wouldn't hold my breath on that, now that it's transitioned to WIP status. (I mean, I'm not saying it'll never get done, but the initially-unexpected/unplanned backwards-compatibility requirement — something I hadn't even _considered_ when writing it; I came up with the whole implementation on the fly while I way laying it out in that comment — makes it a significantly bigger change than it started out. Which means I want to take the time to get it right, rather than rushing it in.)

I meant to say the same to @DylanC regarding #2276, which he'd previously (in #2275 I think?) indicated was sort of "on hold" pending #2312. Now that #2312 has no concrete timeframe, and has expanded into its own bigger thing, I'd say don't worry about that... in fact, it may be _better_ if any other changes related to recovery/thumbnails get merged **before** I finish #2312, and I can account for them rather than vice versa.

**Comment by ferdnyc** [comment]
> Btw, some inside knowledge from Peter on the Undo steps. If you do a split clip and "keep both sides" it will have 3 undo steps. This should be investigated at some point to see could we have less undo steps for this task.

And ISTR is _literally_ three undo steps, meaning if you use it you have to hit Undo three times to get back to where you started.

That's... just lame, TBH, but from an undo stack perspective it shouldn't matter, since those three steps will (or _should_!) count as 3 of the max 20 entries in the stack. So, I don't believe it's any kind of factor in the growth of the undo data / project files.

**Comment by ferdnyc** [comment]
> My main editor is Sublime Text 3 but funnily enough the screenshot is just from an online json parser.  http://json.parser.online.fr/

Ooh! Bookmarked. Their beta interface is rather nice, too, even if the loss of the `[+]` / `[-]` buttons makes folding a little less obvious. (You just click directly on curly or square brackets to fold.) Still, the turning on "Show Indexes" gives you _much_ more visible and obvious ordinals than before, and means you don't _have_ to fold in order to count:
![image](https://user-images.githubusercontent.com/538020/48003548-c5289c00-e0dc-11e8-8d2e-55a5a9fcb6c5.png)

> One of the issues with the history is that it stores the entire before/after state of the project data for each change.

So, I take this back. Each history entry actually includes only the entire _old_ state of the affected object in `old_values`, plus a `"key"` pointing to where on which list those values came from, and a `"value"` object with some summary details about the changed object. 

So, the history's size _doesn't_ grow with the project complexity (each individual object of a given type always has the same size state, no matter how many of them there are), and its size _should_ be fixed at 
<code>_length_ × history_steps</code>.

All of which is to say, there's something very, **VERY** wrong with a project file that contains 11.8MB of history out of 12MB total size, and please _please_ say you still have that version of the file and can share it somewhere, @peanutbutterandcrackers . This is starting to sound like a very severe bug.

**Comment by N3WWN** [comment]
See also: https://github.com/OpenShot/openshot-qt/issues/1837#issuecomment-404880119

😉

**Comment by ferdnyc** [comment]
...However, all that being said, `"clip"` objects are by far the largest stored in the JSON data, so the fact that most history entries involve Clip changes makes the history still rather data-hungry even under ideal circumstances. Especially since all of the _unchanged_ values are still duplicated at each step, and they store their contents somewhat inefficiently to begin with.

(By that I mean, e.g...)
```json
               "location_x" : {
                  "Points" : [
                     {
                        "co" : {
                           "Y" : 0,
                           "X" : 1
                        },
                        "interpolation" : 2
                     }
                  ]
               },
               "location_y" : {
                  "Points" : [
                     {
                        "interpolation" : 2,
                        "co" : {
                           "Y" : 0,
                           "X" : 1
                        }
                     }
                  ]
               },
```
...rather than...
```json
               "location_x" : [
                     { "co": [1, 0],
                       "interpolation" : 2
                     },
                  ]
// or
               "location_y" : [
                     {
                        "frame" : 1,
                        "value" : 0,
                        "interpolation" : 2,
                     },
               ]
```
...or even...
```json
               "location" : [
                      {
                        "frame" : 1,
                        "XY" : [0, 0],
                        "interpolation" : 2
                     },
               ]
```

Since that's what each X-Y `"co"`-ordinate pair on the `"Points"` list **really** represents: a frame number (`"X"`) and a value (`"Y"`) for a single scalar parameter, the only type really supported. Even the color keyframes get broken down into _four_ scalar parameters (`"red"`, `"green"`, `"blue"`, and `"alpha"`), each with their own separate interpolation and bézier-curve parameters — even though I don't believe there's any support for actually _shaping_ the different channel curves independently, nor any real need to add that support.

The "X"/"Y" designation unfortunately obscure the reality of the `"co"`-ordinates' meaning, and reinforces the presumption that _actual_ two-dimensional coordinates (like `"location"`) or tuples (like a 4-dimensional RGBA `"color"`) need to be split into separate `"location_x"` / `"location_y"`-type scalars, instead of being treated as multi-dimensional. The keyframes' "fake" `"co"`-ordinates are probably also making it harder (at least conceptually) to switch from frame-based to time-based keyframe positioning, the critical problem at the core of the long-standing unchangeable framerate bug.

A more efficient data format would make project-data handling lighter. As efficient as JSON is in general, OpenShot pushes that efficiency to the absolute limit rather than really benefiting from it. 

(_Not_ repeating default values over and over again would probably also help, though there'd certainly have to be a section of the project file that _stored_ the local defaults, in case they later changed in the code. Still, for a parameter like e.g. `"interpolation"`, where the value is `2` well over 90% of the time, storing that only _once_ in a `"defaults": { "interpolation" : 2, }` object, and then storing only the rare `"interpolation": 3` or `"interpolation": 37` as part of a keyframe where it was modified, can't have _zero_ efficiency benefit.)

**Comment by ferdnyc** [comment]
> See also: [#1837 (comment)](https://github.com/OpenShot/openshot-qt/issues/1837#issuecomment-404880119)

(GitHub really needs a "screams in abject horror!" reaction...)

**Holy.** **_BALLS!_**

**Comment by peanutbutterandcrackers** [comment]
> Yeah, I wouldn't hold my breath on that

So... [there you go](https://github.com/peanutbutterandcrackers/meet-tux). I saved the project using your backwards-incompatible branch. So, that is the only way to test the project out without messing the thumbnails now. The old 12 MB `.osp` file resides in `Debug` directory.

There's also a sed script in `Scripts` to delete history from a `.osp` project.

Hope that helps!

![meet tux](https://user-images.githubusercontent.com/26764547/48016787-a1a13800-e154-11e8-9f1d-bfe2a64a6ad4.gif)

**Comment by ferdnyc** [comment]
> So... [there you go](https://github.com/peanutbutterandcrackers/meet-tux).  [...] The old 12 MB `.osp` file resides in `Debug` directory.

It took `json_pp` (running on my 4-core, 8MB, Core i5-2400 system with filesystems on SSD — not a speed-demon, but it does OK) **TWENTY-ONE SECONDS** to pretty-print the contents of that file. The resulting output is 99MB, and contains _**1.4 million**_ lines of text.

I'm basically afraid to even open it, TBH.

> I saved the project using your backwards-incompatible branch. So, that is the only way to test the project out without messing the thumbnails now. 

Well, _that's_ even more easily solved than going the other direction:
```console
$ perl -p -i.bak -e 's{\@thumbnail/}{/home/username/.openshot_qt/thumbnail/}g;' filename.osp
```
And, done.

(The same could be done with `sed`, of course. But I like Perl's `-i.suffix` backup functionality, and making (safe) in-place file modifications. Plus, the `s{}{}` syntax allows me to be totally lazy about escaping most special characters.)

**Comment by ferdnyc** [comment]
Hoo-boy. Okay, yeah, you can see from the line numbers of the (folded) history entries exactly where things started to go wrong:

![image](https://user-images.githubusercontent.com/538020/48020309-a4277180-e103-11e8-8d97-c64195f8d8e5.png)

The _interesting_ thing is that there are actually only **three** entries that exploded. 

They are (counting forward): 
* the 4th (lines 312-321310)
* the 18th (lines 329985-877844)
* the 19th (lines 877845-1425704)

All of the _other_ history entries, including the most recent one, were a normal size of up to "a few hundred" lines. These normal-sized entries occurred even _after_ the first of the explodey ones.

So the question will be, what's so special about history entries 4, 18, and 19, that they completely destroyed this project file? And does this mean OpenShot's in-memory copy of the project data was _also_ corrupted, before it was written to disk? Or is this purely an error in the process of marshaling the project state variables for JSON export, and wouldn't impact OpenShot's in-memory representation if not for the fact that it's later going to be read into memory from the corrupted file?

**Comment by ferdnyc** [comment]
And here's the contents of history entry 4:

![image](https://user-images.githubusercontent.com/538020/48021196-f9648280-e105-11e8-8752-5583857ca021.png)

Some observations:
1. `"old_values"` contains a `null`. 
   (Which may be valid, in certain circumstances. Say the history entry were the creation of a new Clip object. There'd be no _old_ state for that object.) 
1. The `"key"`, which in the previous entry is `[ "clips", { "id" : "CLIPID" }, "effects", { "id" : "EFFECTID" } ]` and is used to store a pointer to the part of the project data being updated, is actually set to `[ "history" ]`, indicating that the entire history list, both undo and redo, is being `"update"`-d by this history entry.
1. As a result (and _in a sense_ "correctly", given the other data so far), the `"values"` object therefore contains a copy of **an entire history stack**.
1. The history stack copied into history entry 4, based on the line numbers of its own folded members, _also_ contains some exploded entries. In each of those, I'm sure we'll find yet another `[ "history" ]` type update, containing another full history stack, and so on and so on ala the recursive insanity from @N3WWN 's referenced #1837.

So, long and short of it: 

### Somehow, at certain times and under certain conditions, modifications to the history stack are ending up as entries _ON_ the history stack. 

This is extremely, extremely bad.

**Comment by ferdnyc** [comment]
> * The history stack copied into history entry 4, based on the line numbers of its own folded members, _also_ contains some exploded entries. In each of those, I'm sure we'll find yet another `[ "history" ]` type update, containing another full history stack, and so on and so on ala the recursive insanity from @N3WWN 's referenced #1837.

Having recursively examined its members, in _just_ that 4th history entry alone (the first one that really exploded the file), there are 9 entire history stacks, at various levels up to 3 deep. 

Obviously, the innermost ones are fairly small. (Only 2000-3000 lines, barely more than 3-5x the size of a regular data update.) But repeat the whole mess a couple of times recursively, times multiple siblings in the same state, and it starts to add up to real money.

**Comment by DylanC** [comment]
@ferdnyc - Thanks for your great detective work here. I think from what I've been reading in this thread this issue should be our one and only priority right now. It may be the issue causing 90% of our performance issues/sluggish lag with the user interface.

Now that we have an idea of what is happening the next thing we need to tackle is finding the **exact** cause which is more tricky. We know that simple project changes do not really create this scenario, this scenario is more experienced by users who create more complex content. 

I guess keeping an eye on the history content while editing a project is probably the way to narrow down what is causing this?

**Comment by ferdnyc** [comment]
> Now that we have an idea of what is happening the next thing we need to tackle is finding the **exact** cause which is more tricky. We know that simple project changes do not really create this scenario, this scenario is more experienced by users who create more complex content.
> 
> I guess keeping an eye on the history content while editing a project is probably the way to narrow down what is causing this?

Maybe it's time to add some debug tooling to OpenShot, for use when monitoring issues like this. In _theory_ the terminal/log could be used — just throw some additional messages in there, to spit out whatever data we want to monitor. Problem is, the console output is _already_ so oversaturated with  debugging messages that got left in the code, using it to trace a *new* issue like this one would require first going through and _removing_ all of the existing logspam. (TBH, the logs might _already_ contain all the info needed to track down this bug, but good luck slogging through them to find it. Be easier to just walk the _code_ looking for it.)

Instead, I'm thinking maybe an additional dock, only present in development versions (like, defined in a source file that only lives in the source tree, and gets excluded from release or AppImage packaging), and activated by launching OpenShot in the tree via `src/debug.py` instead of `src/launch.py`.

The Debug Dock could be home to whatever instrumentation we'd build to monitor issues like this, but (in my thinking) display a rolling "dashboard" type snapshot of interesting/useful metrics, rather than a spammy log stream. The first indicator: a total node count for the `_data` object, refreshed every N seconds. That's all, nothing more. Should be enough to make "explosion" events noticeable, which hopefully would be enough to start closing in on both the triggering factors for the bug, and the location of its root cause within the code.

I'd be up for contributing (to) an effort like that, if there's interest. Can't promise how quickly it'd progress, but co-conspirators would be welcome to help divide up the work and keep it moving...

**Comment by ferdnyc** [comment]
(Oh, and yes, whatever data the "dashboard" displays should **also** be logged to the console, for purposes of tracing / reconstructing event timelines.) 

...Just as long as the messages are clear and identifiable, so  they're easily grepped for and contain sufficient context. IOW, as long as they don't end up being more of this kind of thing:
```
 main_window:INFO keyPressEvent: Right
      logger:INFO 16777236
 main_window:INFO keyPressEvent: Right
      logger:INFO 16777236
 main_window:INFO keyPressEvent: Right
      logger:INFO 16777236
 main_window:INFO keyPressEvent: Right
      logger:INFO 16777236
 main_window:INFO keyPressEvent: Right
      logger:INFO 16777236
timeline_webview:INFO SortItems
timeline_webview:INFO UpdateLayerIndex
 main_window:INFO updateStatusChanged
preview_thread:INFO refreshFrame
preview_thread:INFO self.player.Position(): 2377
preview_thread:INFO onModeChanged
preview_thread:INFO onModeChanged
```

**Comment by ferdnyc** [comment]
Yeah, so I think I got as far into this as copying `src/launch.py` to `src/debug.py` and editing exactly 0 lines of code in it. The only reason it isn't identical to `src/launch.py` is that it doesn't have my recent commit to fix the commandline argument handling. All of which is to say, I've made zero progress, for lack of time.

Given that, and my suspicion that large spans of free time will not be magically appearing in my near future, at least, I would propose that our focus for addressing this issue, at least in the short term, should be limited to just getting the history _out_ of all project files, in all cases.

If the history's not saved, it can't be saved in "exploded" form even when it **does** explode (for reasons not yet known), and this way users who are having history-related performance problems _at the very least_ will have the option to save their work, restart OpenShot, and load their project back in. That should **always** clear up performance issues related to the size of the history stack, as long as the project file isn't accompanied by its potentially-massive history information.

(We should probably also change the project-file _loader_ to always discard any and all history data in older project files, when loading those. That way, this will also solve the problem for people whose existing project files **already** contain exploded history stacks.)

If at some point the history issues are sorted out, and the history becomes safe to save and reload, then that functionality can be added back in. (Though it would also require addressing the AppImage session-specific paths thing, too.)

What say you all?
